codec-http2/src/main/java/io/netty/handler/codec/http2/PriorityStreamByteDistributor.java

```diff
@@ -0,0 +1,428 @@
+                // Grow the array by a factor of 2.
+                stillHungry = Arrays.copyOf(stillHungry, min(maxSize, stillHungry.length << 1));
+            }
+        }
+    }
+
+    /**
+     * A simplified version of {@link ChildFeeder} that is only used when all streamable bytes fit
+     * within the available connection window.
+     */
+    private final class SimpleChildFeeder implements Http2StreamVisitor {
+        int bytesAllocated;
+        int connectionWindow;
+
+        SimpleChildFeeder(int connectionWindow) {
+            this.connectionWindow = connectionWindow;
+        }
+
+        @Override
+        public boolean visit(Http2Stream child) {
+            PriorityState childState = state(child);
+            int bytesForChild = childState.unallocatedStreamableBytes();
+
+            if (bytesForChild > 0 || childState.hasFrame()) {
+                childState.allocate(bytesForChild);
+                bytesAllocated += bytesForChild;
+                connectionWindow -= bytesForChild;
+            }
+            int childBytesAllocated = allocateBytesForTree(child, connectionWindow);
+            bytesAllocated += childBytesAllocated;
+            connectionWindow -= childBytesAllocated;
+            return true;
+        }
+    }
+
+    /**
+     * The remote flow control state for a single stream.
+     */
+    private final class PriorityState {
+        final Http2Stream stream;
+        boolean hasFrame;
+        int streamableBytes;
+        int allocated;
+        int unallocatedStreamableBytesForTree;
+
+        PriorityState(Http2Stream stream) {
+            this.stream = stream;
+        }
+
+        /**
+         * Recursively increments the {@link #unallocatedStreamableBytesForTree()} for this branch in
+         * the priority tree starting at the current node.
+         */
+        void unallocatedStreamableBytesForTreeChanged(int delta) {
+            unallocatedStreamableBytesForTree += delta;
+            if (!stream.isRoot()) {
+                state(stream.parent()).unallocatedStreamableBytesForTreeChanged(delta);
+            }
+        }
+
+        void allocate(int bytes) {
+            allocated += bytes;
+
+            // Also artificially reduce the streamable bytes for this tree to give the appearance
+            // that the data has been written. This will be restored before the allocated bytes are
+            // actually written.
+            unallocatedStreamableBytesForTreeChanged(-bytes);
+        }
+
+        /**
+         * Reset the number of bytes that have been allocated to this stream by the priority
+         * algorithm.
+         */
+        void resetAllocated() {
+            allocate(-allocated);
+        }
+
+        void updateStreamableBytes(int newStreamableBytes, boolean hasFrame) {
+            this.hasFrame = hasFrame;
+
+            int delta = newStreamableBytes - streamableBytes;
+            streamableBytes = newStreamableBytes;
```

Comment:
@nmittler - Couldn't `streamableBytes = newStreamableBytes;` also go int he scope of the `delta != 0` block?